### DISCUSSION TITLE
Switch case in C.

### DISCUSSION THREAD

**Discussion by anzosasuke** [NationalSecurityAgency/ghidra#4372] [Q&A] [answered]
I am trying to find the switch case using ghidra. I was wondering if ghidra has an API that could help me to identify switch case in 'ghidra headless'. I see that decompiled output can identify switch case. I would like to do it using headless. Could anyone point me to the right direction? I saw `ghidra.util.state.analysis.switch` but didn't see any information.

```
switch (digit) { //Writing a case for every digit in 0-9
 
   case 1:
           printf("One\n");
           break;
   case 2:
           printf("Two\n");
           break;
   case 3:
           printf("Three\n");
           break;
   case 4:
           printf("Four\n");
           break;
```

Some of the marker of switch cases are indirect jumps to the Jump table. Is there any way to get the jumptable or findout the existence of the jump table?

**Comment by gipi** [comment]
I saw you posted the same question on reddit where I tried to answer, I post the answer here to make other people find it and maybe partecipate to the discussion.

For the ``Switch`` class I'm sorry to inform you that is not usable directly because it's an abstract class and the static method is not callable from outside the package itself.

There is only one class that is calling that method and it's [MySwitchAnalyzer](https://github.com/NationalSecurityAgency/ghidra/blob/da94eb86bd2b89c8b0ab9bd89e9f0dc5a3157055/Ghidra/Features/Base/src/main/java/ghidra/util/state/analysis/MySwitchAnalyzer.java) but bad enough it's not working for me, also because I don't know what it should do.

An approach would be to rewrite the logic inside the [Switch](https://github.com/NationalSecurityAgency/ghidra/blob/da94eb86bd2b89c8b0ab9bd89e9f0dc5a3157055/Ghidra/Features/Base/src/main/java/ghidra/util/state/analysis/Switch.java#L51) class (and I'm not even sure would do what you are asking for).

Otherwise you can parse the C code from the decompiler, using this snippet

```
options = DecompileOptions()

ifc = DecompInterface()
ifc.setOptions(options)
ifc.openProgram(currentProgram)

monitor = ConsoleTaskMonitor()
res = ifc.decompileFunction(caller, 60, monitor)

markup = res.getCCodeMarkup()
```

you will obtain a [ClangTokenGroup](https://ghidra.re/ghidra_docs/api/ghidra/app/decompiler/ClangTokenGroup.html) instance and maybe start a painful journey.

**Comment by gipi** [comment, marked as answer]
After a little more exploring I find out that exists the method ``getJumpTables()`` from ``HighFunction``

```
>>> from ghidra.app.decompiler import DecompileOptions
>>> from ghidra.app.decompiler import DecompInterface
>>> from ghidra.util.task import ConsoleTaskMonitor
>>> def decompile_function(func):
>>>     options = DecompileOptions()
>>>     monitor = ConsoleTaskMonitor()
>>>     ifc = DecompInterface()
>>>     ifc.setOptions(options)
>>>     ifc.openProgram(getCurrentProgram())
...     return ifc.decompileFunction(func, 60, monitor)
>>> res = decompile_function(getFunctionContaining(currentAddress))
>>> tables = res.getHighFunction().getJumpTables()
>>> len(tables)
1
>>> table = tables[0]
>>> table.getSwitchAddress()
0015a33c
>>> table.getLabelValues()
array(java.lang.Integer, [0, 1, 2, 3, 4, 5, 6])
```

How this is hidden in the documentation is a mistery.

**Reply by anzosasuke** [reply]
Thanks @gipi I was able to get the switch table. I also saw the getLoadTables() in the ghidra. I was trying but wasn't able to use that. any idea how?

`tables[0].getLoadTables()[0].getSize()`

Any idea what loadtable is ?

**Reply by gipi** [reply]
no, I should take a look at the source code, however if you are interested to the addresses of the jump table I would take a look at the ``getCases()`` of the ``table`` or analyse the blocks from the ``Pcode`` point of view

```
>>> get_block_at = lambda addr: [_ for _ in res.getHighFunction().getBasicBlocks() if _.contains(addr)][0]
>>> [get_block_at(_) for _ in table.getCases()]
[basic@0015a344, basic@0015a348, basic@0015a34c, basic@0015a350, basic@0015a354, basic@0015a358, basic@0015a35c]
```

or instruction point of view

```
>>> from ghidra.util.task import TaskMonitor
>>> [bm.getCodeBlocksContaining(_, TaskMonitor.DUMMY) for _ in table.getCases()]
[array(ghidra.program.model.block.CodeBlock, [caseD_0  src:[0015a33c] dst:[0015a43c]]), array(ghidra.program.model.block.CodeBlock, [caseD_1 src:[0015a33c]  dst:[0015a4a0]]), array(ghidra.program.model.block.CodeBlock, [caseD_2  src:[0015a33c]  dst:[0015a44c]]),
array(ghidra.program.model.block.CodeBlock, [caseD_3  src:[0015a33c] dst:[0015a45c]]), array(ghidra.program.model.block.CodeBlock, [caseD_4 src:[0015a33c]  dst:[0015a46c]]), array(ghidra.program.model.block.CodeBlock, [caseD_5  src:[0015a33c]  dst:[0015a478]]),
array(ghidra.program.model.block.CodeBlock, [caseD_6  src:[0015a33c] dst:[0015a484]])]
```

**Reply by anzosasuke** [reply]
Thanks a lot @gipi
